python/lsst/daf/butler/core/progress.py

```diff
@@ -0,0 +1,396 @@
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = ("Progress", "ProgressBar", "ProgressHandler")
+
+from abc import ABC, abstractmethod
+from contextlib import contextmanager
+import logging
+from typing import (
+    ClassVar,
+    Collection,
+    ContextManager,
+    Generator,
+    Iterable,
+    Iterator,
+    Optional,
+    Protocol,
+    Tuple,
+    TypeVar,
+)
+
+_T = TypeVar("_T", covariant=True)
+_K = TypeVar("_K")
+_V = TypeVar("_V", bound=Collection)
+
+
+class ProgressBar(Protocol, Iterable[_T]):
+    """A structural interface for progress bars that wrap iterables.
+
+    An object conforming to this interface can be obtained from the
+    `Progress.bar` method.
+
+    Notes
+    -----
+    This interface is intentionally defined as the intersection of the progress
+    bar objects returned by the ``click`` and ``tqdm`` packages, allowing those
+    to directly satisfy code that uses this interface.
+    """
+
+    def update(self, n: int = 1) -> None:
+        """Increment the progress bar by the given amount.
+
+        Parameters
+        ----------
+        n : `int`, optional
+            Increment the progress bar by this many steps (defaults to ``1``).
+            Note that this is a relative increment, not an absolute progress
+            value.
+        """
+        pass
+
+
+class Progress:
+    """Public interface for reporting incremental progress in the butler and
+    related tools.
+
+    This class automatically creates progress bars (or not) depending on
+    whether a handle (see `ProgressHandler`) has been installed and the given
+    name and level are enabled.  When progress reporting is not enabled, it
+    returns dummy objects that can be used just like progress bars by calling
+    code.
+
+    Parameters
+    ----------
+    name : `str`
+        Name of the process whose progress is being reported.  This is in
+        general the name of a group of progress bars, not necessarily as single
+        one, and it should have the same form as a logger names: the
+        fully-qualified name of the class, method, or function, minus the
+        initial "lsst" (and without any "implemntation detail" packages or
```

Comment:
Minus the inital "lsst" if outside of daf_butler....

Reply:
I'll just remove the parenthetical.